lib/jsonapi/utils/string.ex

```diff
@@ -110,6 +110,58 @@ defmodule JSONAPI.Utils.String do
     end
   end
 
+  @doc """
+  Replace underscores between words in `value` with dashes
+
+  Ignores underscores that are not between letters/numbers
+
+  ## Examples
+
+      iex> camelize("top_posts")
+      "topPosts"
+
+      iex> camelize("_top_posts")
+      "_topPosts"
+
+      iex> camelize("_top__posts_")
+      "_top__posts_"
+  """
+  def camelize(value) when is_atom(value) do
+    value
+    |> to_string()
+    |> camelize()
+  end
+
+  def camelize(value) when is_binary(value) do
+    case Regex.split(~r{([a-zA-Z0-9])(?<delimeter>[-_])([a-zA-Z0-9])}, to_string(value), on: [:delimeter]) do
+      words ->
+        [h | t] = words |> Enum.filter(&(&1 != ""))
```

Comment:
```elixir
  def camelize(value) when is_binary(value) do
    [h | t] =
      ~r/([a-zA-Z0-9])(?<delimeter>[-_])([a-zA-Z0-9])/
        |> Regex.split(value, on: [:delimeter], trim: true)
        |> Enum.map(&String.capitalize/1)

    Enum.join([String.downcase(h) | t])
  end
end
```

Reply:
I updated my example above to use the `length(t) > 0` guard because I found an issue with the original code (and mine) using this regex:

```elixir
Utils.String.camelize("exampleName")
"examplename"
```

I do not believe we want to downcase strings that are already camelcased.

Reply:
We can also update our regex here to avoid the need for the `on: [:delimeter]` (delimiter is spelled wrong in the regex also):

```elixir
~r/(?<=[a-zA-Z0-9])[-_](?=[a-zA-Z0-9])/

iex> Regex.split(~r/(?<=[a-zA-Z0-9])[-_](?=[a-zA-Z0-9])/, "example_name")
["example", "name"]
```

Reply:
@snewcomer / @jherdman I feel this file as gotten to be a bit complicated with lots of functions for each style of formatting.  What do you think about simplifying this module a bit?

```elixir
defmodule Example do
  def underscore(value) do
    value
    |> tokenize()
    |> join("_")
  end

  def dasherize(value) do
    value
    |> tokenize()
    |> join("-")
  end

  def camelize(value) do
    with [h | t] when length(t) > 0 <- tokenize(value),
         capitalized <- Enum.map(t, &String.capitalize/1),
         downcased <- String.downcase(h)
    do
      Enum.join([downcased | capitalized])
    else
      _ -> value
    end
  end

  defp join(value, joiner), do: Enum.join(value, joiner)

  defp tokenize(value) do
    replaced_value = String.replace(value, ~r/([a-z])(?=[A-Z])/, "\\1_\\2")
    tokens = Regex.split(~r/(?<=[a-zA-Z0-9])[-_](?=[a-zA-Z0-9])/, replaced_value, trim: true)
    Enum.map(tokens, &String.downcase/1)
  end
end
```

```shell
iex> Example.underscore("example_name")
"example_name"
iex> Example.underscore("example-name")
"example_name"
iex> Example.underscore("exampleName")
"example_name"

iex> Example.dasherize("example-name")
"example-name"
iex> Example.dasherize("example_name")
"example-name"
iex> Example.dasherize("exampleName")
"example-name"

iex> Example.camelize("example_name")
"exampleName"
iex> Example.camelize("example-name")
"exampleName"
iex> Example.camelize("exampleName")
"exampleName"
```

Happy to open a PR to clean up this util 👍